src/app/clusters/bridged-device-basic-information-server/bridged-device-basic-information-server.cpp

```diff
@@ -0,0 +1,280 @@
+/**
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ *
+ */
+
+#include "bridged-device-basic-information-server.h"
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/EventLogging.h>
+#include <app/util/attribute-storage.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConfigurationManager.h>
+#include <platform/PlatformManager.h>
+
+#include <cstddef>
+#include <cstring>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::BridgedDeviceBasic;
+using namespace chip::app::Clusters::BridgedDeviceBasic::Attributes;
+using namespace chip::DeviceLayer;
+
+namespace {
+
+class BridgedDeviceBasicInformationAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the Bridged Device Basic Information cluster on all endpoints.
+    BridgedDeviceBasicInformationAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), BridgedDeviceBasic::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+};
+
+BridgedDeviceBasicInformationAttrAccess gAttrAccess;
+
+CHIP_ERROR EncodeStringOnSuccess(CHIP_ERROR status, AttributeValueEncoder & encoder, const char * buf, size_t maxBufSize)
+{
+    ReturnErrorOnFailure(status);
+    return encoder.Encode(CharSpan(buf, strnlen(buf, maxBufSize)));
+}
+
+CHIP_ERROR BridgedDeviceBasicInformationAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    if (aPath.mClusterId != BridgedDeviceBasic::Id)
+    {
+        // We shouldn't have been called at all.
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    CHIP_ERROR status = CHIP_NO_ERROR;
+
+    switch (aPath.mAttributeId)
+    {
+    case VendorName::Id: {
+        constexpr size_t kMaxLen     = DeviceLayer::ConfigurationManager::kMaxVendorNameLength;
+        char vendorName[kMaxLen + 1] = { 0 };
+        status                       =
+            DeviceLayer::PlatformMgr().GetBridgedVendorName(aPath.mEndpointId, vendorName, sizeof(vendorName));
```

Comment:
It doesn't really make sense to ask the platform manager for this info.

On an actual bridge, the way this would work is that the endpoint would be dynamic and the bridge itself would implement the attributes either via AttributeAccessInterface or via just marking them external and handling the callbacks.  I don't think it makes any sense to have a "non-bridge" implementation of this cluster for all-clusters-app...

Reply:
Is there documentation on the architecture of attribute access? I'm not sure if I fully understand it. Would handling it via AttributeAccessInterface just mean using the generated accessors? And how would it handle the callbacks differently if they stay marked external?

You mean removing it entirely from all-clusters-app? Is there a reason it was already in there, as well as some other clusters?

Reply:
There is some documentation at https://github.com/project-chip/connectedhomeip/blob/master/src/app/AttributeAccessInterface.h#L367-L385 but not anything that describes the overall set of options for attribute storage...  I'll see if I can write up something.

But in brief, there are four options:

1. Attribute handled entirely by the managed attribute store.  Reads and writes are handled automatically, the attribute is stored in RAM and will also be written through to NVM if the "NVM" option is selected in ZAP.  Writes to the attribute call notification functions that the cluster and the application can implement to find out when the value of the attribute changed.  This option is only available for static endpoints that the code generator knows about, so can allocate space for them in the attribute store.
2. Attribute handled by an AttributeAccessInterface implemented in the SDK itself.  This is used to handle attributes where we want to run some logic on every read, or for attributes that the generic attribute store does not know how to store (lists and structs).  If the attribute is not marked as "External" storage, this can fall back to option 1 as described at the link above.
3. Attribute handled by an AttributeAccessInterface implemented in the application.  This makes sense when the SDK itself could not very easily provide the values.  This situation could _also_ be handled by doing option 2 and having the SDK implementation talk to the application via some defined delegate interface, which is more or less what this PR is doing, right?  But if the only logic in the SDK is glue code, why bother?
4. Attribute handled by application external attribute callbacks.  These get called for attributes with "External" storage that are not handled by an AttributeAccessInterface.  It's not clear to me that we really need this option, since all consumers of it can in fact use an AttributeAccessInterface, but so far we have not gotten rid of it (it came with the original Silicon Labs data model code).

I assume this cluster is in all-clusters-app because (1) someone took the name literally and (2) people wanted to run YAML tests against this cluster.  But since this cluster really only makes sense for dynamic endpoints (bridged ones), what we should really do is remove it from all-clusters-app, implement in some sane way in bridge-app, and make sure that we can run some subset of the YAML tests against bridge-app.

Reply:
Thanks for the clarifications!

Reply:
@bzbarsky-apple 

I agree VendorName attribute should be implemented on bridge device instead of the node communicating with a Bridge(bridged device), shall we review the following attributes defined in the spec for the node communicating with a Bridge.

DataModelRevision | X
VendorName | O
VendorID | O
ProductName | O
ProductID | X
NodeLabel | O
Location | X
HardwareVersion | O
HardwareVersionString | O
SoftwareVersion | O
SoftwareVersionString | O
ManufacturingDate | O
PartNumber | O
ProductURL | O
ProductLabel | O
SerialNumber | O
LocalConfigDisabled | X
Reachable | M
UniqueID | O

The spec does not say which attributes should reside on the bridge device or the bridged device itself

Reply:
We need to review which attributes shall be implemented on the bridge device or the bridged device (Node communicating with the bridge), the spec does not say it clearly.

DataModelRevision | X
VendorName | O
VendorID | O
ProductName | O
ProductID | X
NodeLabel | O
Location | X
HardwareVersion | O
HardwareVersionString | O
SoftwareVersion | O
SoftwareVersionString | O
ManufacturingDate | O
PartNumber | O
ProductURL | O
ProductLabel | O
SerialNumber | O
LocalConfigDisabled | X
Reachable | M
UniqueID | O

Reply:
The bridge needs to implement these attributes (modulo optionality).  Note that the only mandatory attribute here is "Reachable".

Whether it implements them by querying the bridged device every time, or by caching information about it, or by having the user manually input the information once and saving it is up to the bridge app implementation.

And the one required thing (Reachable) clearly needs to be implemented by the bridge itself.

Reply:
I was intending to submit this in a separate PR. Can this PR go in as-is?